Fatalization of goto LABEL construct

jkeenan opened the pull request:
This pull request supersedes https://github.com/Perl/perl5/pull/23677.  (I am filing a new pull request in part because in my own GH repo I changed the name of the branch I wish to merge into blead; network search suggests that filing a new p.r. is the easiest way to proceed.)  This commit message incorporates aspects of the earlier p.r.'s commit message.

I've been working on this project over the past month-and-a-half. I am now ready to propose a plan to implement the fatalization of the behavior currently guarded by warnings category `deprecated::goto_construct`.  This pull request would resolve https://github.com/Perl/perl5/issues/23618.

This project is, in effect, a **feature removal** process, so we have to move carefully.  If we remove a feature from core, we also have to remove code testing or otherwise relying on that feature.  We have to remove documentation about that feature except insofar as we discuss the removal or suggest other ways to approach the business problem the feature was previously intended to solve.  We then have to assess how the feature removal would affect Perl code out in the wild -- which may, in turn, require developing new programs (outside of the core distribution) to facilitate that assessment.  I have done all that, so this pull request is being submitted in non-draft status with the hope that we can get it into blead well before the Feb 20 2026 (v5.43-8) "contentious changes" deadline for our next production release in May 2026.

On September 1 we completed one stage in this project. In [#23659](https://github.com/Perl/perl5/pull/23659) ([c00b151](https://github.com/Perl/perl5/commit/c00b151b1cdbf6ce617e50c8a2ec073adcf85f5b)), we moved many of the code blocks in `t/op/goto.t` which we anticipate will *not* be affected by this fatalization to new file `t/op/goto-sub.t`.

The remaining stages then became:

* **Preliminary code changes in the Perl guts**, including removal of the existing deprecation warning and its replacement with a new exception.  To support the new exception we'll probably have to update documentation in `pod/perldeprecation.pod` and `perldoc -f goto`.  We should get `make test_porting` to PASS.  We should identify tests in the test suite *other than* `t/op/goto*.t` which need to test for an exception rather than a warning.  All of this should, IMO, be done *before* we tackle `t/op/goto.t` head-on.  With the exception of needing more detail in the entry for the fatalization in `pod/perldiag.pod`, all of this has been done in this pull request.

* **Getting the main test files to PASS.**  That file is `t/op/goto.t`, which now passes largely because I have simply stripped out all the code that pertained to the now fatal feature.  The extent to which the code remaining in that file exercises ("covers") the remaining `goto` features needs investigation.

* Identifying the scope of **Blead Breaks CPAN (BBC)** problems.  Since CPANtesters is set up on the premise that we're only testing CPAN distros against perl production/maintenance releases or commits to `blead`, we'll have to figure out a way to test against a branch without polluting the CPANtesters matrix.  I have done that by revising one of my previous CPAN distributions and releasing it as [Test-Against-Commit](https://metacpan.org/dist/Test-Against-Commit).  I'll include details of my research using that library in a separate post.

Please review.  Thank you very much.

-------------------------------------------------------------------------------
* This set of changes requires a perldelta entry, and I need help writing it.

tonycoz commented on t/uni/labels.t:
```diff
@@ -75,12 +75,17 @@ like $@, qr/Unrecognized character/, "redo to downgradeable labels";
-        ここ: {
-            pass("goto UTF-8 LABEL works.");
+    local $@;
+    eval {
+        goto ここ;
+
+        if (undef) {
+            ここ: {
+               my $x = "jump goto UTF-8 LABEL no longer works";
+            }
         }
-    }
+    };
+    like($@,
+        qr/Use of goto to jump into a construct is no longer permitted/,
+        "Got expected error message");
```
From what I can see the purpose of this test is to check whether we can goto a unicode label.

While it still checks that the label is resolved, it no longer checks whether we can goto a unicode label.

jkeenan replied:
In b003602b7d, I added a test block which I believe does what you are asking for.  Agreed?

jkeenan replied:
@tonycoz ^^

tonycoz commented on pp_ctl.c:
```diff
@@ -3652,9 +3652,7 @@ PP(pp_goto)
-                deprecate_fatal_in(WARN_DEPRECATED__GOTO_CONSTRUCT,
-                        "5.42",
-                        "Use of \"goto\" to jump into a construct");
+                croak("Use of goto to jump into a construct is no longer permitted");
```
Other similar messages "quote" the operator name, I think this should continue to do so.

jkeenan replied:
> Other similar messages "quote" the operator name, I think this should continue to do so.

I was going to fight you on this, because I'm starting to loathe single and double quotes and apostrophes in error messages.  When I want to copy highlighted copy from a terminal onto a command-line, I can never remember what the escaping rules are, inside and outside of regexes.  And let's not get started on Unix vs DOS.  Moreover, I could cite many messages in `pod/perldiag.pod` that do very well without and single- or double-quoting of keywords.

However, some mostly unrelated research into the history of this deprecation showed that the double-quoted `"goto"` has been around in various forms for many years.  Hence, restoring in https://github.com/Perl/perl5/pull/23782/commits/ae4d500a5516b0523fe0e084cd11ccba46165f1c.

tonycoz commented on t/porting/deprecation.t:
```diff
@@ -90,56 +90,61 @@ if (-e ".git") {
+#    do "regen.pl"; # this should produce a deprecation warning
+#    like($warning, qr/is no longer in \@INC/,
+#        "no warnings 'deprecated::goto_construct'; did not silence deprecated::dot_in_inc warnings");
+#}
```
The commented out blocks above should be adapted or removed.

Since the only remaining deprecation I see in pod/perldeprecation is scheduled to be fatal in 5.42, I expect it should be removed.

jkeenan replied:
Done in 656f8d5d13

tonycoz commented on t/comp/package_block.t:
```diff
@@ -81,12 +81,7 @@ eval q{
     }
     $main::result .= "j(".__PACKAGE__."/".eval("__PACKAGE__").")";
 };
-print $main::result eq
-	"a(main/main)d(Foo/Foo)g(main/main)i(Bar/Bar)j(main/main)" ?
-    "ok 6\n" : "not ok 6\n";
-print $main::warning =~ /\A
-	Use\ of\ "goto"\ [^\n]*\ line\ 3\.\n
-	Use\ of\ "goto"\ [^\n]*\ line\ 15\.\n
-    \z/x ? "ok 7\n" : "not ok 7\n";
+print $main::result eq "a(main/main)" ?  "ok 6\n" : "not ok 6\n";
+print $main::warning eq '' ? "ok 7\n" : "not ok 7\n";
```
Rather than capturing the new result with the gotos croaking (truncating the result) I think it should change to not use block package syntax.

jkeenan replied:
I'm not grokking what you're saying.  Do you mean that everything between lines 61 and 83 should be rewritten?

tonycoz replied:
The `goto l0` and `goto l2` are now fatal due to this change, since they're jumping into the package {} blocks.

Instead of using block form packages these could change to inline packages, ie from:

```
package Foo {
    ...
}
...
package Bar {
    ...
}
```

to:
```
package Foo;
...
package Bar;
...
```
and still test that goto is properly setting the visible `__PACKAGE__` which is the intent of the test.

jkeenan replied:
This test file is called `package_block.t`.  IIUC, if we make the changes you are recommending, they would no longer be tests of the package block syntax.  Where should they go instead?

jkeenan replied:
I moved those two tests to `t/op/goto.t` and reworked them to use `t/test.pl`.  Pushed in 9729f844f9; please review.

tonycoz commented on t/lib/croak/pp_ctl:
```diff
@@ -1,23 +1,20 @@
 CORE::given(goto f){f:}
 EXPECT
-Can't "goto" into a "given" block at - line 2.
+Use of goto to jump into a construct is no longer permitted at - line 1.
```
It would be nice if we could keep the more detailed messages here, though that would take some figuring out.

jkeenan replied:
In blead, those other exception messages are defined in pp_ctl.c in the `S_check_op_type` function:
```
3229 S_check_op_type(pTHX_ OP * const o)
3230 {
3231     /* Eventually we may want to stack the needed arguments
3232      * for each op.  For now, we punt on the hard ones. */
3233     /* XXX This comment seems to me like wishful thinking.  --sprout */
3234     if (o == UNENTERABLE)
3235         croak(
3236                   "Can't \"goto\" into a binary or list expression");
3237     if (o->op_type == OP_ENTERITER)
3238         croak(
3239                   "Can't \"goto\" into the middle of a foreach loop");
3240     if (o->op_type == OP_ENTERGIVEN)
3241         croak(
3242                   "Can't \"goto\" into a \"given\" block");
3243 }
3244 
```
Note the inline comments and misgivings in the above -- all from b5377742950 by Father C back in 2017.

The warning about impending fatalization occurs later on in the same file, in `PP(pp_goto)`:
```
3247 PP(pp_goto)
3248 {
...
3649         if (*enterops && enterops[1]) {
3650             I32 i = enterops[1] != UNENTERABLE
3651                  && enterops[1]->op_type == OP_ENTER && in_block
3652                     ? 2
3653                     : 1;
3654             if (enterops[i])
3655                 deprecate_fatal_in(WARN_DEPRECATED__GOTO_CONSTRUCT,
3656                         "5.42",
3657                         "Use of \"goto\" to jump into a construct");
3658         }
...
3705 }
```
In blead, we see these messages here:
```
$ ack 'into the middle of a foreach loop|into a binary or list expression|into a.+?given.+?block' *.c t/
pp_ctl.c
3236:                  "Can't \"goto\" into a binary or list expression");
3239:                  "Can't \"goto\" into the middle of a foreach loop");
3242:                  "Can't \"goto\" into a \"given\" block");

t/lib/croak/pp_ctl
7:Can't "goto" into the middle of a foreach loop at - line 3.
14:Can't "goto" into a "given" block at - line 3.
20:Can't "goto" into a "given" block at - line 2.
29:Can't "goto" into a binary or list expression at - line 2.
30:Can't "goto" into a binary or list expression at - line 3.
31:Can't "goto" into a binary or list expression at - line 4.
32:Can't "goto" into a binary or list expression at - line 5.

t/op/goto.t
286:    like($@, qr/Can't "goto" into the middle of a foreach loop/,

t/op/eval.t
228:	like($@, qr/Can't "goto" into the middle of a foreach loop/,
```
Now I acknowledge that I don't understand why, in blead, the deprecation warning is sometimes encountered first, whereas other times 1 of the 3 `croak` messages is encountered first.

In the branch, we see them here:
```
$ ack 'into the middle of a foreach loop|into a binary or list expression|into a.+?given.+?block' *.c t/
pp_ctl.c
3236:                  "Can't \"goto\" into a binary or list expression");
3239:                  "Can't \"goto\" into the middle of a foreach loop");
3242:                  "Can't \"goto\" into a \"given\" block");

t/lib/croak/pp_ctl
26:Can't "goto" into a binary or list expression at - line 2.
27:Can't "goto" into a binary or list expression at - line 3.
28:Can't "goto" into a binary or list expression at - line 4.
29:Can't "goto" into a binary or list expression at - line 5.

t/op/eval.t
228:	like($@, qr/Can't "goto" into the middle of a foreach loop/,
```
The 3 messages are still defined.  2 of the 3 are still exercised in either `t/lib/croak.t` or `t/op/eval.t`.  The message about a `given` block is no longer exercised.  If the point of this fatalization is to have an exception throw *whenever* we jump to a LABEL that is located inside a "construct" (a poorly defined term, IMO), then I don't see why we need the 3 more specific messages at all anymore.  If you want to retain them and test for them, you should make an argument for that (though that would constitute mission creep, IMO).

But, as I acknowledged in earlier posts, I've never used `goto` myself, so I could be missing something.

tonycoz commented:
I spent some time looking at this, removing the `leaving_eval` condition from the conditional around line 3643 improves the error messages, but I'm not sure that's the correct change at this point.

tonycoz commented:
> All tests PASSed -- which suggests that the error messages either did not change or are not being exercised. Can you describe the improvements you see?

I was working with the f31afb2 version of the PR at the time:
```
t/lib/croak ... FILE: lib/croak/pp_ctl ; line 2
PROG: 
goto f;
foreach(1){f:}
EXPECTED:
Use of "goto" to jump into a construct is no longer permitted at - line 1.
EXIT STATUS: != 0
GOT:
Can't "goto" into the middle of a foreach loop at - line 1.
EXIT STATUS: 255
# Failed test 95 - goto into foreach at lib/croak/pp_ctl line 2
FILE: lib/croak/pp_ctl ; line 8
PROG: 
goto f;
CORE::given(1){f:}
EXPECTED:
Use of "goto" to jump into a construct is no longer permitted at - line 1.
EXIT STATUS: != 0
GOT:
Can't "goto" into a "given" block at - line 1.
EXIT STATUS: 255
# Failed test 96 - goto into given at lib/croak/pp_ctl line 8
FILE: lib/croak/pp_ctl ; line 14
PROG: 
CORE::given(goto f){f:}
EXPECTED:
Use of "goto" to jump into a construct is no longer permitted at - line 1.
EXIT STATUS: != 0
GOT:
Can't "goto" into a "given" block at - line 1.
EXIT STATUS: 255

```

jkeenan committed:
Fatalize use of goto to jump into construct

* Adapt all tests.

* Revise discussion of 'goto' in perlfunc.

* Perl 5.44 fatalization: document goto_construct Also, note that no
  deprecations or fatalizations occurred in perl-5.42.

* Remove 'goto_construct' from list of warnings.  This entails:

** Entering 'deprecated::goto_construct" into %NO_BIT_FOR inside
regen/warnings.pl;

** Incrementing $VERSION in that file;

** Running `make regen` to propagate these changes to lib/warnings.pm and
warnings.h, in the process renumbering the warnings categories and
adding 'deprecated::goto_construct' to %NoOp inside lib/warnings.pm.

* Add entry to perldiag.pod

* Update perldelta: partial fatalization of goto construct.

* t/op/goto.t: Restore tests deleted during work on branch.

tonycoz commented on pod/perldiag.pod:
```diff
@@ -7923,6 +7923,10 @@ For speed and efficiency reasons, Perl internally does not do full
 
+=item Use of "goto" to jump into a construct is no longer permitted
+
+(F) More TO COME.
```
This looks incomplete

jkeenan replied:
Documentation fleshed out in 1311220206.

tonycoz commented on t/op/goto.t:
```diff
@@ -379,24 +253,22 @@ sub DEBUG_TIME() {
-    no_list:
-        is($out, 'perl rules', '$out has not been erroneously reset to undef');
+        no_list:
+            is($out, 'perl rules', '$out has not been erroneously reset to undef');
```
why was this re-indented?

jkeenan replied:
Should be restored in 1dbcd43f12.

jkeenan commented:
On 12/9/25 17:58, Tony Cook wrote:
> ***@***.**** commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In pod/perldiag.pod <https://github.com/Perl/perl5/ 
> pull/23782#discussion_r2604634235>:
> 
>> @@ -7923,6 +7923,10 @@ For speed and efficiency reasons, Perl internally does not do full
>   reference-counting of iterated items, hence deleting such an item in the
>   middle of an iteration causes Perl to see a freed value.
>   
> +=item Use of "goto" to jump into a construct is no longer permitted
> +
> +(F) More TO COME.
> 
> This looks incomplete
> 

Sorry, I thought I had completed that.  I will look into this tomorrow. 
Any other current concerns?

leonerd reviewed (approved):
LGTM

leonerd reviewed (approved):
LGTM
